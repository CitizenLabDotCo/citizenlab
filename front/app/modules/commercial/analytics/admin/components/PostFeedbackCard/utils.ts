import { CornerRadius } from 'components/admin/Graphs/typings';

export const getCornerRadius =
  (stackLength: number, cornerRadius: number) =>
  ({ stackIndex }: { stackIndex: number }): CornerRadius => {
    const r = cornerRadius;
    if (stackIndex === 0) return [r, 0, 0, r];
    if (stackIndex === stackLength - 1) return [0, r, r, 0];

    return 0;
  };
