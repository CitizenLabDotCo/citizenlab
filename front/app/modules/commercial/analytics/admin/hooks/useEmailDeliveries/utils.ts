import { TimeSeriesResponse } from './typings';
import { IResolution } from 'components/admin/ResolutionControl';

export const deduceResolution = (
  timeSeriesResponse: TimeSeriesResponse
): IResolution | null => {
  if (timeSeriesResponse.length === 0) return null;
  const firstRow = timeSeriesResponse[0];

  if ('dimension_date_last_action.month' in firstRow) {
    return 'month';
  }

  if ('dimension_date_last_action.week' in firstRow) {
    return 'week';
  }

  return 'day';
};
