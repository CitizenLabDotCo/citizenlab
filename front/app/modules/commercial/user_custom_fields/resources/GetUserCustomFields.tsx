import React from 'react';
import { Subscription } from 'rxjs';
import {
  IUserCustomFieldData,
  IUserCustomFieldInputType,
  userCustomFieldsStream,
} from '../services/userCustomFields';

interface InputProps {}

type children = (
  renderProps: GetUserCustomFieldsChildProps
) => JSX.Element | null;

interface Props extends InputProps {
  children?: children;
  inputTypes?: IUserCustomFieldInputType[];
}

interface State {
  userCustomFields: IUserCustomFieldData[] | undefined | null;
}

export type GetUserCustomFieldsChildProps =
  | IUserCustomFieldData[]
  | undefined
  | null;

export default class GetCustomFields extends React.Component<Props, State> {
  private subscriptions: Subscription[];

  constructor(props: Props) {
    super(props);
    this.state = {
      userCustomFields: undefined,
    };
  }

  componentDidMount() {
    const { inputTypes } = this.props;
    const userCustomFields$ = userCustomFieldsStream({
      queryParameters: { input_types: inputTypes },
    }).observable;

    this.subscriptions = [
      userCustomFields$.subscribe((userCustomFields) => {
        this.setState({
          userCustomFields: userCustomFields.data,
        });
      }),
    ];
  }

  componentWillUnmount() {
    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
  }

  render() {
    const { children } = this.props;
    const { userCustomFields } = this.state;
    return (children as children)(userCustomFields);
  }
}
