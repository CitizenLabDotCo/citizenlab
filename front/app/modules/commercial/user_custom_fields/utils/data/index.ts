export { default as binBirthyear } from './binBirthyear';
export { default as join } from './join';
export { default as rename } from './rename';
export {
  default as convertDomicileData,
  AreaValue,
} from './convertDomicileData';
export * from './typings';
