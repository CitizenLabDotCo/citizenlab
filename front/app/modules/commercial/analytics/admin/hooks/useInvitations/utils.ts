import { WrappedComponentProps } from 'react-intl';
import messages from '../../components/InvitationsCard/messages';
import { getTimePeriodTranslationByResolution } from '../../utils/resolution';

export interface InvitationsLabels {
  cardTitle: string;
  fileName: string;
  total: string;
  pending: string;
  accepted: string;
  period: string;
}

export const getInvitationsLabels = (
  formatMessage: WrappedComponentProps['intl']['formatMessage'],
  resolution
): InvitationsLabels => {
  return {
    cardTitle: formatMessage(messages.invitations),
    fileName: formatMessage(messages.invitations)
      .toLowerCase()
      .replace(' ', '_'),
    total: formatMessage(messages.totalInvites),
    pending: formatMessage(messages.pending),
    accepted: formatMessage(messages.accepted),
    period: getTimePeriodTranslationByResolution(formatMessage, resolution),
  };
};
