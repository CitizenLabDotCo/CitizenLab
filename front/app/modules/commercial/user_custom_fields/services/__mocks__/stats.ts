import { BehaviorSubject } from 'rxjs';
import { IStreamParams } from 'utils/streams';
import { IUsersByRegistrationField } from '../stats';

let mockUsersByRegFieldsVariable: IUsersByRegistrationField | null = null;

export const __setMockUsersByRegFields = (data: IUsersByRegistrationField) => {
  mockUsersByRegFieldsVariable = data;
};

export const usersByRegFieldStream = jest.fn((_streamParams: IStreamParams, _regFieldId) => {
  const observable = new BehaviorSubject(mockUsersByRegFieldsVariable);
  return {
    observable
  };
});

export const mockUsersByRegFields = {
  series: {
    users: {
      active_politician: 1,
      no: 7,
      retired_politician: 2,
      _blank: 11
    }
  },
  options: {
    active_politician: {
      title_multiloc: {
        en: 'Active politician'
      }
    },
    retired_politician: {
      title_multiloc: {
        en: 'Retired politician'
      }
    },
    no: {
      title_multiloc: {
        en: 'No'
      }
    }
  }
} as IUsersByRegistrationField;
