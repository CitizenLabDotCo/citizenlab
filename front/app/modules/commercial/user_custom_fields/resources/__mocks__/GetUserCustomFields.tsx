import { GetUserCustomFieldsChildProps } from '../../resources/GetUserCustomFields';

export const mockGetUserCustomFields = [
  {
    id: '3d9ad03c-fbe3-47e5-8d6d-c8bbc512682d',
    type: 'custom_field',
    attributes: {
      key: 'gender',
      input_type: 'select',
      title_multiloc: {
        en: 'Gender',
        'da-DK': 'Gender',
        'de-DE': 'Geschlecht',
        'en-CA': 'Gender',
        'en-GB': 'Gender',
        'fr-BE': 'Genre',
        'fr-FR': 'Genre',
        'nb-NO': 'Gender',
        'nl-BE': 'Geslacht',
        'nl-NL': 'Geslacht',
      },
      description_multiloc: {},
      required: false,
      ordering: 0,
      enabled: true,
      code: 'gender',
      created_at: '2018-12-03T11:01:46.392Z',
      updated_at: '2018-12-03T11:01:46.392Z',
    },
  },
  {
    id: '45f7f17d-5b05-4be4-879a-ca9a040596c0',
    type: 'custom_field',
    attributes: {
      key: 'domicile',
      input_type: 'select',
      title_multiloc: {
        en: 'Place of residence',
        'da-DK': 'Place of residence',
        'de-DE': 'Wohnort',
        'en-CA': 'Place of residence',
        'en-GB': 'Place of residence',
        'fr-BE': 'Domicile',
        'fr-FR': 'Domicile',
        'nb-NO': 'Place of residence',
        'nl-BE': 'Woonplaats',
        'nl-NL': 'Woonplaats',
      },
      description_multiloc: {},
      required: false,
      ordering: 2,
      enabled: true,
      code: 'domicile',
      created_at: '2018-12-03T11:01:46.436Z',
      updated_at: '2018-12-03T11:01:46.436Z',
    },
  },
  {
    id: '4145ae1d-da1e-4984-a043-e1269eb37cd4',
    type: 'custom_field',
    attributes: {
      key: 'multiselect_test',
      input_type: 'multiselect',
      title_multiloc: {
        en: 'Multiselect test ',
        'nl-BE': 'Multiselect test ',
      },
      description_multiloc: {
        en: 'Multiselect test field',
        'nl-BE': 'Multiselect test field',
      },
      required: false,
      ordering: 5,
      enabled: true,
      code: null,
      created_at: '2018-12-03T11:16:19.452Z',
      updated_at: '2018-12-03T11:16:19.452Z',
    },
  },
  {
    id: 'a0e5a642-a67d-4f73-bb52-35af018a903c',
    type: 'custom_field',
    attributes: {
      key: 'are_you_ready',
      input_type: 'checkbox',
      title_multiloc: {
        en: 'Are you ready ?',
      },
      description_multiloc: {
        en: 'Test checkbox',
      },
      required: false,
      ordering: 6,
      enabled: true,
      code: null,
      created_at: '2018-12-03T11:18:06.161Z',
      updated_at: '2018-12-03T11:18:06.161Z',
    },
  },
] as GetUserCustomFieldsChildProps;
