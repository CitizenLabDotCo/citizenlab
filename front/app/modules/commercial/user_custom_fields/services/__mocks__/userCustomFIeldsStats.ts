import { IUsersByRegistrationField } from '../stats';

export const mockUsersByRegFields = {
  series: {
    users: {
      active_politician: 1,
      no: 7,
      retired_politician: 2,
      _blank: 11,
    },
  },
  options: {
    active_politician: {
      title_multiloc: {
        en: 'Active politician',
      },
    },
    retired_politician: {
      title_multiloc: {
        en: 'Retired politician',
      },
    },
    no: {
      title_multiloc: {
        en: 'No',
      },
    },
  },
} as IUsersByRegistrationField;
