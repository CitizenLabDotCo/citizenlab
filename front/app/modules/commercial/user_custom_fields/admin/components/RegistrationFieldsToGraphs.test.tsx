// @ts-nocheck
import React from 'react';
import { shallow } from 'enzyme';
jest.mock('utils/cl-intl');
jest.mock('../../resources/GetUserCustomFields');
jest.mock('services/stats');
jest.mock('components/Outlet', () => 'Outlet');
jest.mock('modules', () => ({ streamsToReset: [] }));

import {
  RegistrationFieldsToGraphs,
  Props,
} from './RegistrationFieldsToGraphs';
import { mockGetUserCustomFields } from '../../resources/__mocks__/GetUserCustomFields';
import { mockUsersByRegFields } from '../../services/__mocks__/userCustomFIeldsStats';
import { localizeProps } from 'utils/testUtils/localizeProps';
import { intl } from 'utils/cl-intl';

describe('<RegistrationFieldsToGraphs />', () => {
  // this is a unit test to verify that this component "parses" the props that were passed in, and
  // renders an array of different components depending on the data with props like filters and time
  it('renders correctly', () => {
    const wrapper = shallow(
      <RegistrationFieldsToGraphs
        startAt="someTimeAgo"
        endAt="now"
        currentGroupFilter={null}
        customFields={mockGetUserCustomFields}
        intl={intl}
        {...localizeProps}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('correcly handles data', () => {
    const wrapper = shallow<Props>(
      <RegistrationFieldsToGraphs
        startAt="someTimeAgo"
        endAt="now"
        currentGroupFilter={null}
        customFields={mockGetUserCustomFields}
        intl={intl}
        {...localizeProps}
      />
    );
    const convertToGraphFormat = wrapper
      .find('WrappedBarChartByCategory')
      .prop('convertToGraphFormat');
    expect(convertToGraphFormat(mockUsersByRegFields)).toMatchSnapshot();
  });
});
