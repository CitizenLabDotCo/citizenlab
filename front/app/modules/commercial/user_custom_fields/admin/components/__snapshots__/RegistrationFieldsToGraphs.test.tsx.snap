// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegistrationFieldsToGraphs /> correcly handles data 1`] = `
Array [
  Object {
    "code": "active_politician",
    "name": "Active politician",
    "value": 1,
  },
  Object {
    "code": "retired_politician",
    "name": "Retired politician",
    "value": 2,
  },
  Object {
    "code": "no",
    "name": "No",
    "value": 7,
  },
  Object {
    "code": "_blank",
    "name": "unknown",
    "value": 11,
  },
]
`;

exports[`<RegistrationFieldsToGraphs /> renders correctly 1`] = `
Array [
  <WithTheme(injectIntl(GenderChart))
    currentGroupFilter={null}
    endAt="now"
    key="0"
    startAt="someTimeAgo"
  />,
  <injectIntl(WithLocalize(AreaChart))
    currentGroupFilter={null}
    endAt="now"
    key="1"
    startAt="someTimeAgo"
  />,
  <WrappedBarChartByCategory
    convertToGraphFormat={[Function]}
    currentGroupFilter={null}
    customId="4145ae1d-da1e-4984-a043-e1269eb37cd4"
    endAt="now"
    graphTitleString="Multiselect test "
    graphUnit="users"
    key="2"
    startAt="someTimeAgo"
    stream={[Function]}
    xlsxEndpoint="/web_api/v1/stats/users_by_custom_field_as_xlsx/4145ae1d-da1e-4984-a043-e1269eb37cd4"
  />,
  <WrappedPieChartByCategory
    convertToGraphFormat={[Function]}
    currentGroupFilter={null}
    customId="a0e5a642-a67d-4f73-bb52-35af018a903c"
    endAt="now"
    graphTitleString="Are you ready ?"
    graphUnit="users"
    key="3"
    startAt="someTimeAgo"
    stream={[Function]}
    xlsxEndpoint="/web_api/v1/stats/users_by_custom_field_as_xlsx/a0e5a642-a67d-4f73-bb52-35af018a903c"
  />,
]
`;
